Remove python_cmake_module from ros2.repos.

clalancette opened the pull request:
This PR aims to remove `python_cmake_module` from the core ROS 2 repositories.

The basic idea behind this is that `python_cmake_module` is an unnecessary indirection to `find_package(Python3)`, which is a standard CMake feature.  So we may as well remove it and save that additional layer of indirection.

While we are doing this, we also notice that we can support venvs in ROS 2 much better by taking advantage of some newer features of `find_package(Python3)`.  In particular, we can set [CMP0094](https://cmake.org/cmake/help/latest/policy/CMP0094.html), which sets `Python3_FIND_STRATEGY` to `LOCATION`.  In short, this means that `find_package(Python3)` will stop searching after it finds the first python3 binary it can.

However, `find_package(Python3)`, at least on Linux, will attempt to find the most specific version it can by default.  That means that if both `python3.11` and `python3` are on the PATH, it will by default choose `python3.11`.  We can also control *that* behavior by setting [Python3_FIND_UNVERSIONED_NAMES](https://cmake.org/cmake/help/latest/module/FindPython3.html#hints) to `FIRST`.

The end result of all of this is that venvs on Linux should work properly, as long as the venv is activated.  On Windows, due to different logic, this always worked properly.  If all else fails, the user can always pass `--cmake-args -DPython3_EXECUTABLE=/path/to/python3`, which will forcibly set the path to the python executable.

Given the complexity of Python packaging, I don't expect it will fix all related problems, but it should improve many of our venv issues, like https://github.com/ros2/ros2cli/issues/879 , https://github.com/ament/ament_cmake/issues/502 , https://github.com/ros2/ros2/issues/1519 , https://github.com/ros2/ros2/issues/1094 , and https://github.com/ros2/ros2/issues/1430 .

Also note that this is *not* backportable to Humble, since it requires CMake 3.20 and Humble has to support CMake versions older than that.  There we are using a simpler version of this, in https://github.com/ros2/geometry2/pull/650 and https://github.com/ament/ament_cmake/pull/507 .  This might be backportable to Iron, but we'll have to see.

Note that this is a draft because we need to get all of the dependent PRs in first, and I still have to write up the documentation and CI code for this.

yashi commented:
Just wanted to say that if you have `libpython3.12-dev` installed but your `python3` is `3.11.8`, the following two find different versions.  This is the current situation on Debian Sid right now.

- `find_package(Python3 REQUIRED COMPONENTS Development)`
- `find_package(Python3 REQUIRED COMPONENTS Interpreter)`

This leads to a build failure on some ROS related packages even with this PR applied.

https://gist.github.com/yashi/0661b04106661e75b1749d3f057331e7

yashi commented:
Just wanted to say that if you have `libpython3.12-dev` installed but your `python3` is `3.11.8`, the following two find different versions.  This is the current situation on Debian Sid right now.

- `find_package(Python3 REQUIRED COMPONENTS Development)`
- `find_package(Python3 REQUIRED COMPONENTS Interpreter)`

This leads to a build failure on some ROS related packages even with this PR applied.

https://gist.github.com/yashi/0661b04106661e75b1749d3f057331e7

clalancette commented:
> Just wanted to say that if you have `libpython3.12-dev` installed but your `python3` is `3.11.8`, the following two find different versions. This is the current situation on Debian Sid right now.

Yeah, there is just nothing we can do about that.  That is a bug in the distribution.

clalancette commented:
All right, I've now updated this series.  Here's is CI again on all of this, including Windows Debug:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=21226)](http://ci.ros2.org/job/ci_linux/21226/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=15536)](http://ci.ros2.org/job/ci_linux-aarch64/15536/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=700)](http://ci.ros2.org/job/ci_linux-rhel/700/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=21947)](http://ci.ros2.org/job/ci_windows/21947/)
* Windows Debug [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=21948)](http://ci.ros2.org/job/ci_windows/21948/)

clalancette commented:
I've updated the series again, and added in some more fixes for Windows Debug.  Here's another Windows Debug run:

* Windows Debug [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=22062)](http://ci.ros2.org/job/ci_windows/22062/)

clalancette commented:
Updated yet again, here is another CI run:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=21868)](http://ci.ros2.org/job/ci_linux/21868/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=16138)](http://ci.ros2.org/job/ci_linux-aarch64/16138/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1375)](http://ci.ros2.org/job/ci_linux-rhel/1375/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=22580)](http://ci.ros2.org/job/ci_windows/22580/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=22581)](http://ci.ros2.org/job/ci_windows/22581/)
